br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -180,645 +234,570 @@ func physicalTableIDs(tableInfo *model.TableInfo) []int64 {
+	ctx context.Context, keyRange tidbkv.KeyRange,
+) ([]*restore.RegionInfo, []tidbkv.KeyRange, error) {
+	rawStartKey := codec.EncodeBytes(nil, keyRange.StartKey)
+	rawEndKey := codec.EncodeBytes(nil, keyRange.EndKey)
+	allRegions, err := restore.PaginateScanRegion(ctx, m.splitCli, rawStartKey, rawEndKey, 1024)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	regions := make([]*restore.RegionInfo, 0, len(allRegions))
+	keyRanges := make([]tidbkv.KeyRange, 0, len(allRegions))
+	for _, region := range allRegions {
+		_, startKey, err := codec.DecodeBytes(region.Region.StartKey, nil)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		_, endKey, err := codec.DecodeBytes(region.Region.EndKey, nil)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		if bytes.Compare(startKey, keyRange.StartKey) < 0 {
+			startKey = keyRange.StartKey
+		}
+		if bytes.Compare(endKey, keyRange.EndKey) > 0 {
+			endKey = keyRange.EndKey
+		}
+		if bytes.Compare(startKey, endKey) < 0 {
+			regions = append(regions, region)
+			keyRanges = append(keyRanges, tidbkv.KeyRange{
+				StartKey: startKey,
+				EndKey:   endKey,
+			})
+		}
 	}
-	reqs := make([]*DuplicateRequest, 0)
-	for _, r := range keysRanges {
-		req := &DuplicateRequest{
-			start:     r.StartKey,
-			end:       r.EndKey,
-			tableID:   tableID,
-			indexInfo: indexInfo,
+	return regions, keyRanges, nil
+}
+
+// processRemoteDupTask processes a remoteDupTask. A task contains a key range.
+// A key range is associated with multiple regions. processRemoteDupTask tries
+// to collect duplicates from each region.
+func (m *DuplicateManager) processRemoteDupTask(
+	ctx context.Context,
+	task dupTask,
+	logger log.Logger,
+	importClientFactory ImportClientFactory,
+	regionPool *utils.WorkerPool,
+) error {
+	remainKeyRanges := newPendingKeyRanges(task.KeyRange)
+	err := common.Retry("process remote dup task", logger, func() error {
+		var (
+			regions   []*restore.RegionInfo
+			keyRanges []tidbkv.KeyRange
+		)
+		for _, kr := range remainKeyRanges.list() {
+			subRegions, subKeyRanges, err := m.splitKeyRangeByRegions(ctx, kr)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			regions = append(regions, subRegions...)
+			keyRanges = append(keyRanges, subKeyRanges...)
+		}
+
+		var metErr common.OnceError
+		wg := &sync.WaitGroup{}
+		for i := 0; i < len(regions); i++ {
+			region := regions[i]
+			kr := keyRanges[i]
+
+			wg.Add(1)
+			regionPool.Apply(func() {
+				defer wg.Done()
+
+				logger := logger.With(
+					zap.Uint64("regionID", region.Region.Id),
+					logutil.Key("dupDetectStartKey", kr.StartKey),
+					logutil.Key("dupDetectEndKey", kr.EndKey),
+				)
+				err := common.Retry("collect remote duplicate by region", logger, func() error {
+					stream, err := NewRemoteDupKVStream(region, kr, importClientFactory)
+					if err != nil {
+						return errors.Annotatef(err, "failed to create remote duplicate kv stream")
```

Comment:
If the error cause if region epoch change, we can't directly retry without renew the region info.

Reply:
Added to my todo list. Thanks.